akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala

```diff
@@ -0,0 +1,140 @@
+ * [[HttpExt.cachedHostConnectionPool]]. Some other gateways are not shared, such as those obtained through
+ * [[HttpExt.newHostConnectionPool]], and will have their dedicated restartable pool.
+ *
+ */
+private[http] class PoolMasterActor extends Actor {
+
+  import PoolMasterActor._
+
+  private[this] implicit val dispatcher = context.dispatcher
+
+  private[this] var poolStatus = Map[PoolGateway, PoolInterfaceStatus]()
+  private[this] var poolInterfaces = Map[ActorRef, PoolGateway]()
+
+  /**
+   * Start a new pool interface actor, register it in our maps, and watch its death.
+   *
+   * @param gateway the pool gateway this pool corresponds to
+   * @param fm the materializer to use for this pool
+   * @return the newly created actor ref
+   */
+  private[this] def startPoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer): ActorRef = {
+    val props = Props(new PoolInterfaceActor(gateway)).withDeploy(Deploy.local)
+    val ref = context.system.actorOf(props, PoolInterfaceActor.name.next())
+    poolStatus += gateway -> PoolInterfaceRunning(ref)
+    poolInterfaces += ref -> gateway
+    context.watch(ref)
+  }
+
+  def receive = {
+
+    // Start or restart a pool without sending it a request. This is used to ensure that
+    // freshly created pools will be ready to serve requests immediately.
+    case s @ StartPool(gateway, materializer) ⇒
+      poolStatus.get(gateway) match {
+        case Some(PoolInterfaceRunning(_)) ⇒
+        case Some(PoolInterfaceShuttingDown(shutdownCompletedPromise)) ⇒
+          shutdownCompletedPromise.future.foreach(_ ⇒ self ! s)
+        case None ⇒
+          startPoolInterfaceActor(gateway)(materializer)
+      }
+
+    // Send a request to a pool. If needed, the pool will be started or restarted.
+    case s @ SendRequest(gateway, request, responsePromise, materializer) ⇒
+      poolStatus.get(gateway) match {
+        case Some(PoolInterfaceRunning(ref)) ⇒
+          ref ! PoolRequest(request, responsePromise)
+        case Some(PoolInterfaceShuttingDown(shutdownCompletedPromise)) ⇒
+          // The request will be resent when the pool shutdown is complete (the first
+          // request will recreate the pool).
+          shutdownCompletedPromise.future.foreach(_ ⇒ self ! s)
+        case None ⇒
+          startPoolInterfaceActor(gateway)(materializer) ! PoolRequest(request, responsePromise)
+      }
+
+    // Shutdown a pool and signal its termination.
+    case Shutdown(gateway, shutdownCompletedPromise) ⇒
+      poolStatus.get(gateway).foreach {
+        case PoolInterfaceRunning(ref) ⇒
+          ref ! PoolInterfaceActor.Shutdown
+          poolStatus += gateway -> PoolInterfaceShuttingDown(shutdownCompletedPromise)
+        case PoolInterfaceShuttingDown(formerPromise) ⇒
+          shutdownCompletedPromise.completeWith(formerPromise.future)
+        case _ ⇒
+          shutdownCompletedPromise.success(Done)
```

Comment:
I don't see how this would happen. The promise can be completed it three exclusive ways;
- immediately if the gateway does not exist;
- as a completion of an existing promise;
- after being put into a `PoolInterfaceShuttingDown`, and it is only extracted from it in the `Terminated` message.

Or maybe your concern is that an already completed promise could be given to this actor? This should not happen, because the promise is created in the `PoolGateway` which is a private object.